skrub/_total_std_norm.py

```diff
@@ -0,0 +1,108 @@
+import numpy as np
+import sklearn
+from sklearn.utils.fixes import parse_version
+from sklearn.utils.validation import check_array
+
+from . import _dataframe as sbd
+
+
+def total_std_norm(X):
+    r"""Fit the average L2 norm and apply the normalization.
+
+    Compute the average L2 norm (a scalar) from a numerical DataFrame or a 2D NumPy
+    array, and use this norm to normalize the input element-wise. This computation
+    is robust to non-finite values such as ``np.inf`` or ``np.nan``, but raises an
+    error if string values are present.
+
+    Returns
+    -------
+    norm : float,
+        The total standard deviation norm.
+
+    See Also
+    --------
+    :class:`~sklearn.preprocessing.Normalizer` :
+        Performs row-wise normalization.
+
+    Notes
+    -----
+    We define this norm as:
+
+    .. math::
+
+        ||X||_B = \sqrt{\sum_{j=1}^D \sigma^2(X_j)} \in \mathbb{R}
+
+    where:
+
+    * :math:`\sigma^2(X_j)` is the population variance of the column :math:`j`
+    * :math:`D` is the number of features
+
+    When :math:`D = 1`, this norm is the population standard deviation of the column.
+    We then normalize each element as:
+
+    .. math::
+
+        \forall i,j, \quad \tilde{X}_{ij} = \frac{X_{ij}}{||X||_B}
+    """
+
+    if hasattr(X, "__dataframe__"):
+        msg = "X must only have numeric columns, but {col} is {dtype}."
+        for col in sbd.to_column_list(X):
+            if not (sbd.is_numeric(col) or sbd.is_bool(col)):
+                raise ValueError(msg.format(col=col.name, dtype=sbd.dtype(col)))
```

Comment:
in theory it should be `sbd.name(col)` . maybe a little simplification could be to remove the loop and use a selector like `~(s.boolean() | s.numeric())` but not super important

Reply:
BTW, as this is a private function applied to the output of skrub encoders, do we need these checks? I guess if the encoder outputs non-numeric values (which doesn't happen) there's not much the user can do about it anyway?

Reply:
do we need to handle the dataframe case? if this is to be applied to numpy intermediate results inside the encoders, we could say that this function operates on numpy arrays. using it on dataframes should be avoided to avoid the copy into a numpy array

gap, text and string encoders do their computation in numpy, then compute the norm, then rescale and convert to dataframe. doing the rescaling after the first conversion to dataframe would add an unnecessary copy, no?

Reply:
Yes we don't need these extra checks anymore, I initially wrote it to be a public class.